### PR TITLE
Metrics API prototype

### PR DESCRIPTION
**Pull request by lmolkova**
Follow up on https://github.com/Azure/azure-sdk/issues/4389 - show how metrics could be done in .NET before finalizing Java APIs.

### CODE REVIEW

File: sdk/core/Azure.Core/src/Shared/DiagnosticMeter.cs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    internal class DiagnosticMeter : IDisposable
+    {
+        private readonly MetricsOptions? _options;
+        private readonly Meter? _meter;
+
+        public DiagnosticMeter(string name, string? version, MetricsOptions? options)
+        {
+            _meter = new Meter(name, version);
+            _options = options;
+            if (_options == null || _options.IsEnabled)
```
Comment (JoshLove-msft) [inline sdk/core/Azure.Core/src/Shared/DiagnosticMeter.cs:22]: Wouldn't null mean disabled?

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiver.cs [hunk 2]
```diff
@@ -322,11 +340,24 @@ public virtual async Task CloseAsync(CancellationToken cancellationToken = defau
             }
 
             Logger.ReceiveMessageComplete(Identifier, messages);
+            recordLagMetric(messages);
             scope.SetMessageData(messages);
 
             return messages;
         }
 
+        private void recordLagMetric(IReadOnlyList<ServiceBusReceivedMessage> messages)
```
Comment (JoshLove-msft) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiver.cs:349]:
```suggestion
        private void RecordLagMetric(IReadOnlyList<ServiceBusReceivedMessage> messages)
```

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiver.cs [hunk 3]
```diff
@@ -182,6 +190,16 @@ public virtual bool IsClosed
                     // the link is opened.
                     Logger.ClientCreateComplete(type, Identifier);
                 }
+
+                _receiveLag = new DiagnosticMeter("Azure.Messagins.ServiceBus", "7.9.0-beta")
```
Comment (JoshLove-msft) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiver.cs:194]: Do we need to hold onto the DiagnosticMeter reference in order to dispose it?

File: sdk/core/Azure.Core/src/Azure.Core.csproj [hunk 4]
```diff
@@ -8,7 +8,7 @@
     <PackageTags>Microsoft Azure Client Pipeline</PackageTags>
     <Nullable>enable</Nullable>
     <DefineConstants>$(DefineConstants);AZURE_NULLABLE;HAS_INTERNALS_VISIBLE_CORE</DefineConstants>
-    <TargetFrameworks>$(RequiredTargetFrameworks);net461;netcoreapp2.1;net5.0</TargetFrameworks>
+    <TargetFrameworks>$(RequiredTargetFrameworks);net461;net5.0;</TargetFrameworks>
```
[the remark this answers is not in the document]
Comment (lmolkova) [reply sdk/core/Azure.Core/src/Azure.Core.csproj:11]: apparently, DiagnosticSource 6 is not compatible with netcoreapp2.1 - this change won't  be needed as we'd detect DS version in runtime and use reflection

File: sdk/core/Azure.Core/src/Shared/DiagnosticMeter.cs [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        private readonly Histogram<T>? _histogram;
+        internal DiagnosticHistogram(Histogram<T>? histogram)
+        {
+            _histogram = histogram;
+            IsEnabled = _histogram?.Enabled ?? false;
+        }
+
+        public void Record(T value, DiagnosticTags tags)
+        {
+            _histogram?.Record(value, tags.Tags);
```
[the remark this answers is not in the document]
Comment (lmolkova) [reply sdk/core/Azure.Core/src/Shared/DiagnosticMeter.cs:72]: `_histogram` is null when disabled

File: sdk/core/Azure.Core/src/Shared/DiagnosticMeter.cs [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        private readonly Counter<T>? _counter;
+        internal DiagnosticCounter(Counter<T>? counter)
+        {
+            _counter = counter;
+            IsEnabled = _counter?.Enabled ?? false;
+        }
+
+        public void Add(T delta, DiagnosticTags tags)
+        {
+            _counter?.Add(delta, tags.Tags);
```
[the remark this answers is not in the document]
Comment (lmolkova) [reply sdk/core/Azure.Core/src/Shared/DiagnosticMeter.cs:56]: same here, counter is null when disabled

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiver.cs [hunk 7]
```diff
@@ -240,6 +259,7 @@ public virtual async Task CloseAsync(CancellationToken cancellationToken = defau
                 throw;
             }
 
+            _meter.Dispose();
```
Comment (JoshLove-msft) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiver.cs:262]: Should this go in a finally block?

**Review by JoshLove-msft — APPROVED** [review]

**Pull request closed without merging**